### PR TITLE
InSpec tests for 'stackstorm::user'

### PR DESCRIPTION
**Pull request by jvrplmlmn**
Related with: https://github.com/StackStorm/chef-stackstorm/issues/29

- Adds a verifier to KitchenCI (both to Vagrant and Docker).
- Adds InSpec test for 'stackstorm::user'

WDYT @armab @shortdudey123 ?

### CODE REVIEW

**Comment by arm4b** [comment]
You rock! :metal:

File: test/recipes/default/user_test.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+describe user('stanley') do
+  it { should exist }
+  its('group') { should eq 'stanley' }
+  its('home') { should eq '/home/stanley' }
+end
+
+describe passwd.users(/stanley/) do
+  its('homes') { should eq ['/home/stanley'] }
+  its('shells') { should eq ['/bin/bash'] }
```
Comment (arm4b) [inline test/recipes/default/user_test.rb]: Not sure if `/bin/bash` should be a strict requirement. Any edge cases you're thinking of?

File: recipes/user.rb [hunk 2]
```diff
@@ -13,6 +13,7 @@
 
 user conf['user'] do
   home conf.home
+  group conf.group
```
Comment (arm4b) [inline recipes/user.rb]: Good catch! :+1:

Comment (arm4b) [reply recipes/user.rb]: Also can you change the above lines to `conf['home']` and `conf['group']` notation?

I got warnings that it will be deprecated:
> Method access to node attributes (node.foo.bar) is deprecated and will be removed in Chef 13, please use bracket syntax (node["foo"]["bar"])

File: test/recipes/default/user_test.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  it { should be_readable }
+  it { should be_writable.by('owner') }
+  it { should_not be_writable.by('group') }
+  it { should_not be_executable.by('owner') }
+  it { should_not be_executable.by('group') }
+end
+
+describe file('/etc/sudoers.d/stanley') do
+  it { should exist }
+  its('content') { should match(%r{#\s.*?This\s.*?file\s.*?is\s.*?managed\s.*?by\s.*?Chef.}) }
```
Comment (shortdudey123) [inline test/recipes/default/user_test.rb]: `\s.*?`?
Not sure what the purpose of this is for the regex
You should be able to just use a space or `\s`

File: recipes/user.rb (hunk 2 above)
Comment (jvrplmlmn) [reply recipes/user.rb]: Those don't apply since we do `conf = node['stackstorm']['user']`. See:

```
       Deprecated features used!
         method access to node attributes (node.foo.bar) is deprecated and will be removed in Chef 13, please use bracket syntax (node["foo"]["bar"]) at 3 locations:
           - /tmp/kitchen/cache/cookbooks/yum-erlang_solutions/attributes/erlang_solutions.rb:1:in `from_file'
           - /tmp/kitchen/cache/cookbooks/yum-erlang_solutions/attributes/erlang_solutions.rb:2:in `from_file'
           - /tmp/kitchen/cache/cookbooks/mongodb3/recipes/package_repo.rb:26:in `from_file'
         node.set is deprecated and will be removed in Chef 14, please use node.default/node.override (or node.normal only if you really need persistence) at 2 locations:
```

which applies to the `yum-erlang_solutions` and `mongodb3` cookbooks.

Comment (jvrplmlmn) [reply recipes/user.rb]: I'm in favour of getting rid of `conf = node['stackstorm']['user']` and accessing `node['stackstorm']['user'][...]` everywhere in this recipe.

Up to you since this doesn't keep me up at night 😄

File: test/recipes/default/user_test.rb (hunk 1 above)
Comment (jvrplmlmn) [reply test/recipes/default/user_test.rb]: Well I went a little bit overboard with that InSpec resource. `/bin/bash` is the shell that gets assigned to the user when the resource is managed by Chef. If that is not a requirement for StackStorm (maybe for the local action runner?) we can get rid of it.

Comment (arm4b) [reply test/recipes/default/user_test.rb]: Yeah, it's not a requirement for StackStorm. Looks like `bash` is unnecessary information here.

File: test/recipes/default/user_test.rb [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+  its('content') { should match(%r{#\s.*?Do\s.*?NOT\s.*?modify\s.*?this\s.*?file\s.*?directly.}) }
```
Comment (arm4b) [inline test/recipes/default/user_test.rb]: I think we can remove from tests: 
> This file is managed by Chef.
> Do NOT modify this file directly.


The original intent is to verify sudoers entry, doesn't matter if Chef places these comments in file.

**Comment by jvrplmlmn** [comment]
I've addressed all of the comments.

**Review by arm4b — APPROVED** [review]
Nice one! Thanks @jvrplmlmn for starting `InSpec` story!

I think in future it would be fine to Test installations with custom settings: somehow re-use initial attributes passed by user OR another approach, - read `/etc/st2/st2.conf` settings like `[system_user]` and more.

So the Integration Test could be self-sufficient and applied not only just for `stackstorm::bundle` with all default parameters but for custom installations.

In future I expect users to run `InSpec` tests for troubleshooting when they experience any Issues with StackStorm Installation.

**Comment by arm4b** [comment]
On related note, in future I'm thinking about configuring the following directory structure for tests: 

* `test/unit/...` - Chefspec
* `test/integration/...` - Inspec

This will make things more straightforward.
